### PR TITLE
ensure that statics are inhabited

### LINKED ISSUE #74840: Creating a raw reference/pointer to an extern never value claims following code is unreachable
I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3a284533ec584206e289695de0f08df1)):

```rust
#![feature(never_type)]
#![feature(raw_ref_macros)]
#![feature(raw_ref_op)]

extern {
    static FOO: !;
}

fn main() {
    dbg!(unsafe { core::ptr::raw_const!(FOO) });
    dbg!(unsafe { &raw const FOO });
}
```

I expected to see this happen: just a linker failure because `FOO` doesn't exist.

Instead, this happened: before the linker failure, `rustc` complained that code was unreachable.

```
warning: unreachable expression
  --> src/main.rs:10:19
   |
10 |     dbg!(unsafe { core::ptr::raw_const!(FOO) });
   |                   ^^^^^^^^^^^^^^^^^^^^^^---^
   |                   |                     |
   |                   |                     any code following this expression is unreachable
   |                   unreachable expression

warning: unreachable expression
  --> src/main.rs:11:19
   |
11 |     dbg!(unsafe { &raw const FOO });
   |                   ^^^^^^^^^^^---
   |                   |          |
   |                   |          any code following this expression is unreachable
   |                   unreachable expression
```

(The code should also build without the `unsafe`; that is caused by the same problem.)

### Meta
```
1.47.0-nightly
(2020-07-26 6c8927b0cf80ceee1938)
```

cc @rust-lang/wg-unsafe-code-guidelines

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/74840

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_typeck/src/check/check.rs [hunk 1]
```diff
@@ -377,6 +378,36 @@ pub(super) fn check_union_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: Local
     true
 }
 
+/// Check that a `static` is inhabited.
+fn check_static_inhabited<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId, span: Span) {
+    // Make sure statics are inhabited.
+    // Other parts of the compiler assume that there are no uninhabited places. In principle it
+    // would be enugh to check this for `extern` statics, as statics with an initializer will
```
Comment (BlackHoleFox) [inline compiler/rustc_typeck/src/check/check.rs:385]: Small typo: 
```suggestion
    // would be enough to check this for `extern` statics, as statics with an initializer will
```

Comment (RalfJung) [reply compiler/rustc_typeck/src/check/check.rs:385]: Thanks, good catch :)

File: src/test/ui/statics/uninhabited-static.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#![feature(never_type)]
+#![deny(uninhabited_static)]
+
+enum Void {}
+extern {
+    static VOID: Void; //~ ERROR static of uninhabited type
+    //~| WARN: previously accepted
+    static NEVER: !; //~ ERROR static of uninhabited type
```
Comment (oli-obk) [inline src/test/ui/statics/uninhabited-static.rs:8]: does the diagnostic for a regular `static NEVER: ! = panic!();` change with this PR?

Comment (RalfJung) [reply src/test/ui/statics/uninhabited-static.rs:8]: Yes, looks like it does (see the test change I just pushed)... would you prefer it to only apply to foreign items?

Comment (oli-obk) [reply src/test/ui/statics/uninhabited-static.rs:8]: No, having tests for it is good, I'm mostly suprised we don't have any other tests for this

**Pull request merged**